src/Database/Esqueleto/Internal/Internal.hs

```diff
@@ -2172,15 +2172,20 @@ unsafeSqlCase when v = ERaw Never buildCase
     foldHelp _ _ (ECompositeKey _, _) = throw (CompositeKeyErr FoldHelpError)
     foldHelp _ _ (_, ECompositeKey _) = throw (CompositeKeyErr FoldHelpError)
     foldHelp info (b0, vals0) (v1, v2) =
-        let (b1, vals1) = valueToSql v1 info
-            (b2, vals2) = valueToSql v2 info
+        let (b1, vals1) = valueToRawSqlParens v1 info
+            (b2, vals2) = valueToRawSqlParens v2 info
         in ( b0 <> " WHEN " <> b1 <> " THEN " <> b2, vals0 <> vals1 <> vals2 )
 
-    valueToSql :: SqlExpr (Value a) -> IdentInfo -> (TLB.Builder, [PersistValue])
-    valueToSql (ERaw p f) = (first (parensM p)) . f
-    valueToSql (ECompositeKey _) = throw (CompositeKeyErr SqlCaseError)
-    valueToSql (EAliasedValue i _) = aliasedValueIdentToRawSql i
-    valueToSql (EValueReference i i') = valueReferenceToRawSql i i'
+-- | (Internal) Convert a value to a raw SQL builder, preserving parens around
+-- 'ERaw' SQL expressions. This is useful for turning values into function or
+-- operator arguments.
+--
+-- Since: 3.4.0.2
+valueToRawSqlParens :: SqlExpr (Value a) -> IdentInfo -> (TLB.Builder, [PersistValue])
+valueToRawSqlParens (ERaw p f) = (first (parensM p)) . f
+valueToRawSqlParens (ECompositeKey _) = throw (CompositeKeyErr SqlCaseError)
```

Comment:
Just noticed that the error type thrown here is different. Any suggestions?

Reply:
It seems risky to have this same function repeated and specialized more than once in the codebase. I'll go with making the error an argument here.

Reply:
I really don't like how error handling is done in esqueleto at the moment. Since this is a moving of code, it's probably fine to leave as-is. Ideally we'd just fix the representation so this can't happen, but it's not clear to me the best way to do that :thinking: